### ISSUE TITLE
Ivy: after initial build error, takes two builds to emit all files

### ISSUE THREAD

**Issue by filipesilva** [angular/angular#32214]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/compiler-cli

### Is this a regression?

 I don't know

### Description

If an initial build has an error, fixing that error results in two rebuilds instead of one. The first rebuild will only emit the changed files, and the second rebuild will emit the remaining files.

This only happens if there are no emitted files on disk. Using `ngc` this isn't a problem, but it is a problem for the programmatic API consumers like `@ngtools/webpack` that Angular CLI uses.

## 🔬 Minimal Reproduction
- `npm install -g @angular/cli@next`
- `ng new ivy-rebuild-project && cd ivy-rebuild-project`
- update Angular Framework versions to `9.0.0-next.2`
- set `angularCompilerOptions.enableIvy` to `true` in `tsconfig.json` if it's not there
- make sure you don't have the `out-tsc` directory, if you do then delete it
- introduce an AOT error by replacing `selector: 'app-root',` with `selector: (() => 'app-root')(),` in `src/app/app.component.ts`
- run `yarn ngc -p tsconfig.app.json -w`
- wait for the build error message
- remove the AOT error and save the file
- see two incremental builds triggered

## 🔥 Exception or Error
<pre><code>
kamik@RED-X1C6 MINGW64 /d/sandbox/snapshot-project (master)
$ yarn ngc -p tsconfig.app.json -w
yarn run v1.12.3
$ D:\sandbox\snapshot-project\node_modules\.bin\ngc -p tsconfig.app.json -w
src/app/app.component.ts(4,13): error TS-991010: selector must be a string

Compilation failed. Watching for file changes.

src/app/app.component.ts:4:13 - error NG1010: selector must be a string

4   selector: (() => 'app-root')(),
              ~~~~~~~~~~~~~~~~~~~~


File change detected. Starting incremental compilation.

Compilation complete. Watching for file changes.

File change detected. Starting incremental compilation.

Compilation complete. Watching for file changes.
</code></pre>

Note that a single file change yielded two incremental compilations.

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 8.3.0-rc.0
Node: 10.16.0
OS: win32 x64
Angular: 9.0.0-next.2
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.803.0-rc.0
@angular-devkit/build-angular     0.803.0-rc.0
@angular-devkit/build-optimizer   0.803.0-rc.0
@angular-devkit/build-webpack     0.803.0-rc.0
@angular-devkit/core              8.3.0-rc.0
@angular-devkit/schematics        8.3.0-rc.0
@angular/cli                      8.3.0-rc.0
@ngtools/webpack                  8.3.0-rc.0
@schematics/angular               8.3.0-rc.0
@schematics/update                0.803.0-rc.0
rxjs                              6.4.0
typescript                        3.5.3
webpack                           4.39.2
</code></pre>

**Anything else relevant?**
<!-- ✍️Is this a browser specific issue? If so, please specify the browser and version. -->

<!-- ✍️Do any of these matter: operating system, IDE, package manager, HTTP server, ...? If so, please mention it below. -->

**Comment by alxhub** [comment]
I suspect this has something to do with the way we track incremental state after we encounter an error. I will look into it!

**Comment by filipesilva** [comment]
In https://github.com/angular/angular-cli/issues/16036 I am seeing a more severe version of this problem. 

That issue contains a more complex app than the one I mentioned in this repository. To debug what was being emitted, I manually edited the `writeFile` method of the `compilerHost`.

When there's a build error, no files are emitted. Fixing the error emits some of the files, but not all:
```
# /D/sandbox/Angular-Communication/APM-Final v8/src/app/products/product-list.component.js.map
# /D/sandbox/Angular-Communication/APM-Final v8/src/app/products/product-list.component.js
# /D/sandbox/Angular-Communication/APM-Final v8/src/app/products/product-detail.component.js.map
# /D/sandbox/Angular-Communication/APM-Final v8/src/app/products/product-detail.component.js
# /D/sandbox/Angular-Communication/APM-Final v8/src/app/products/edit/product-edit.component.js.map
# /D/sandbox/Angular-Communication/APM-Final v8/src/app/products/edit/product-edit.component.js
# /D/sandbox/Angular-Communication/APM-Final v8/src/app/products/product-shell/product-shell-detail.component.js.map
# /D/sandbox/Angular-Communication/APM-Final v8/src/app/products/product-shell/product-shell-detail.component.js
# /D/sandbox/Angular-Communication/APM-Final v8/src/app/products/product-shell/product-shell-list.component.js.map
# /D/sandbox/Angular-Communication/APM-Final v8/src/app/products/product-shell/product-shell-list.component.js
# /D/sandbox/Angular-Communication/APM-Final v8/src/app/products/product-shell/product-shell.component.js.map
# /D/sandbox/Angular-Communication/APM-Final v8/src/app/products/product-shell/product-shell.component.js
# /D/sandbox/Angular-Communication/APM-Final v8/src/app/products/product.module.js.map
# /D/sandbox/Angular-Communication/APM-Final v8/src/app/products/product.module.js
```

This was a partial emit where the entry point (`main.ts`) wasn't emitted. Even if it had been, a full dependency graph from the entry point down to the emitted files wasn't present. So this rebuild will still fail.

Saving the entry point will only cause it to be emitted, and no other files.
```
# /D/sandbox/Angular-Communication/APM-Final v8/src/main.js.map
# /D/sandbox/Angular-Communication/APM-Final v8/src/main.js
```

**Comment by manughub** [comment]
@vikerman - fyi

**Comment by alxhub** [comment]
I don't think this is actually fixed by the incremental rebuild changes in 4be8929. That commit fixes _some_ cases where errors cause files not to be correctly re-emitted, but I believe it is still possible depending on the change being made.

I'm working on a true fix for this problem.

**Comment by IgorMinar** [comment]
this should be fixed by https://github.com/angular/angular/pull/33971
